test/ambiguous.jl

```diff
@@ -149,21 +147,61 @@ end
 ambs = detect_ambiguities(Ambig5)
 @test length(ambs) == 2
 
+
+using LinearAlgebra, SparseArrays, SuiteSparse
+
 # Test that Core and Base are free of ambiguities
 # not using isempty so this prints more information when it fails
-@test detect_ambiguities(Core, Base; imported=true, recursive=true, ambiguous_bottom=false) == []
-# some ambiguities involving Union{} type parameters are expected, but not required
-@test !isempty(detect_ambiguities(Core, Base; imported=true, ambiguous_bottom=true))
+@testset "detect_ambiguities" begin
+    let ambig = Set{Any}(((m1.sig, m2.sig) for (m1, m2) in detect_ambiguities(Core, Base; recursive=true, ambiguous_bottom=false)))
+        @test_broken isempty(ambig)
+
+        # this is basically just a list of everything for which typeintersection is too conservative,
+        # and ends up matching some methods that don't actually have an interesting intersection
+        expect = []
+        push!(expect, (Tuple{typeof(convert),Type{T},Any} where T>:Union{Missing, Nothing}, Tuple{typeof(convert),Type{T},T} where T>:Nothing))
```

Comment:
I try to understand this list of methods enumerated in the `expect` vector. Since this PR is an improvement, they are previously false-negative and now true-positive, right? If so, is it correct to assume that I should be able to invoke method error with some combination? However, I couldn't get the run-time ambiguity error with a few examples for this first item of `expect`. What am I doing wrong here?

```julia
julia> VERSION
v"1.6.0-DEV.819"

julia> convert(Union{Missing, Nothing, Int}, 1)
1

julia> convert(Union{Missing, Nothing}, missing)
missing

julia> convert(Union{Missing, Nothing}, nothing)

julia> convert(Union{Missing, Nothing, Int}, missing)
missing

julia> convert(Union{Missing, Nothing, Int}, nothing)

julia> convert(Union{Missing, Nothing, Int}, 1)
1
```

Reply:
Not necessarily—the type system might claim that such an ambiguity exists, while no counterfactual actually does. I posted a link above that references this particular failure in the type system (https://github.com/JuliaLang/julia/issues/36951).

Reply:
What is the API of `detect_ambiguities`?  I've been assuming that it returns a non-empty vector if and only if there exist concrete argument types that would cause a method error.  Could you document the exact guarantee?  For example, it would be helpful to document that the users should expect false positive or negative.

If providing a guarantee is not possible or not appropriate (e.g., to allow the type system to evolve), can we demote `detect_ambiguities` (and `detect_unbound_args`?) to experimental API and explicitly mention as such.

Reply:
Aside from problems with the type system, that is what this did before and is still what it does. Hence the new bug report number also. There was a bug in v1.0 where someone got a conditional backwards and accidentally additionally filtered out any ambiguity resulting from a failure of type-intersection. Hence this PR. That's now fixed.

Reply:
> that is what this did before and is still what it does

What is _that_?  Could you explain what `detect_ambiguities` does now?  Could you clarify what kind of "bound" users can assume with respect to the run-time behavior (method error)? No false positive (which is apparently not the case from your previous comment?)? No false negative (i.e., if `detect_ambiguities` returns an empty vector, there will be no method error; given there is no bug, of course)? Both kinds of failures are possible (i.e., it's just a hint)?